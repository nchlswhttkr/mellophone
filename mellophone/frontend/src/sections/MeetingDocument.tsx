import React from "react";

import { IMeeting } from "../types";
import { Link } from "@reach/router";
import Route from "../utils/Route";

interface Props {
  meeting?: IMeeting;
}

export default function MeetingDocument(props: Props) {
  const { meeting } = props;
  if (!meeting) return null;

  return (
    <div>
      <h2>{meeting.name}</h2>
      <h3>
        Meeting held {meeting.dateHeld.toString()} by{" "}
        <Link to={new Route(Route.TEAMS, meeting.team.id).build()}>
          {meeting.team.name}
        </Link>
      </h3>
      <p>
        <em>Yesterday, all my problems seemed so far away...</em>
      </p>
    </div>
  );
}
