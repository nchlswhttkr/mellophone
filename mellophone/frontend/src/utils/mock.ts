import { IUser, ITeam, IMeeting } from "../types";

/**
 * Generates mock domain objects, attempting to populate fields with unique
 * values to avoid accidental collisions and false positives in tests.
 */
class Mock {
  private userCount = 0;
  private teamCount = 0;
  private meetingCount = 0;

  user(user: Partial<IUser> = {}): IUser {
    this.userCount++;
    return {
      id: this.userCount,
      email: `user+${this.userCount}@email.com`,
      firstName: btoa(this.userCount.toString()),
      lastName: btoa(this.userCount.toString()),
      ...user,
    };
  }

  team(team: Partial<ITeam> = {}): ITeam {
    this.teamCount++;
    return {
      id: this.teamCount,
      name: `Team #${this.teamCount}`,
      website: btoa(this.teamCount.toString()),
      ...team,
    };
  }

  meeting(meeting: Partial<IMeeting> = {}): IMeeting {
    this.meetingCount++;
    return {
      id: this.meetingCount,
      name: "Meeting " + btoa(this.meetingCount.toString()),
      venue: "Meeting venue " + btoa(this.meetingCount.toString()),
      dateHeld: new Date(),
      team: this.team(),
      ...meeting,
    };
  }
}

export default new Mock();
