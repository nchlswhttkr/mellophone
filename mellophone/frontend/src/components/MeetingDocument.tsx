import React, { useState, useRef } from "react";

import { IMeeting, IItem } from "../types";
import { Link } from "@reach/router";
import { route } from "../utils/routing";
import Button from "./Button";
import Input from "./Input";
import ErrorMessage from "./ErrorMessage";
import TextArea from "./TextArea";

interface Props {
  meeting: IMeeting;
  items: IItem[];
  createItemForMeeting(meetingId: number, item: Partial<IItem>): Promise<void>;
}

export default function MeetingDocument(props: Props) {
  const [error, setError] = useState<Error>();
  const { meeting, items, createItemForMeeting } = props;

  const itemNameRef = useRef<HTMLInputElement>(null);
  const itemDescriptionRef = useRef<HTMLTextAreaElement>(null);

  function onCreateMeetingItem() {
    const nameInput = itemNameRef.current;
    const descriptionInput = itemDescriptionRef.current;
    if (nameInput && descriptionInput) {
      setError(undefined);
      createItemForMeeting(meeting.id, {
        name: nameInput.value,
        description: descriptionInput.value,
      }).catch(setError);
    }
  }

  return (
    <div>
      <h1>{meeting.name}</h1>
      <h2>
        Meeting held {meeting.dateHeld.toString()} by{" "}
        <Link to={route(`/teams/${meeting.team.id}`)}>{meeting.team.name}</Link>
      </h2>
      <hr></hr>
      {items.map(transformItemIntoElement)}
      <form>
        <Input
          ref={itemNameRef}
          label="Item name"
          aria-label="New meeting item name"
        />
        <TextArea
          ref={itemDescriptionRef}
          label="Item description"
          aria-label="New meeting item description"
          rows={5}
        />
        <Button onClick={onCreateMeetingItem}>Create new item</Button>
        <ErrorMessage error={error} />
      </form>
    </div>
  );
}

// Easier to test this transformation if it is isolation from the component
export const transformItemIntoElement = (item: IItem) => {
  return (
    <div key={item.id}>
      <h3>{item.name}</h3>
      <p>
        {item.description.split("\n").map((line, i) => (
          // Avoids React warning about keys for reconciliation
          <React.Fragment key={i}>
            {i > 0 && <br />}
            {line}
          </React.Fragment>
        ))}
      </p>
    </div>
  );
};
