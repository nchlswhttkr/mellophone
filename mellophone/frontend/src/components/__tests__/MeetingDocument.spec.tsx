import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";

import MeetingDocument, { transformItemIntoElement } from "../MeetingDocument";
import mock from "../../utils/mock";

it("Matches a snapshot from converting an item into an element", () => {
  /**
   * Snapshot testing is useful here to make sure that we break up an item into
   * the expected format, such as <br/> replacing '\n'
   */
  const item = mock.item({
    description: "Line 1\nLine 2\nLine 3",
  });

  const element = transformItemIntoElement(item);

  expect(element).toMatchSnapshot();
});

it("Shows a meeting and its items, and allows new items to be created", () => {
  const meeting = mock.meeting();
  const items = [mock.item()];
  const createItemForMeeting = jest.fn(async () => undefined);

  const { getByLabelText, getByText, queryByText } = render(
    <MeetingDocument
      meeting={meeting}
      items={items}
      createItemForMeeting={createItemForMeeting}
    />
  );

  expect(queryByText(meeting.name)).not.toBe(null);
  // expect(queryByText(meeting.dateHeld)).not.toBe(null)
  expect(queryByText(items[0].name)).not.toBe(null);
  expect(queryByText(items[0].description)).not.toBe(null);

  const newItem = mock.item();

  fireEvent.input(getByLabelText("New meeting item name"), {
    target: { value: newItem.name },
  });
  fireEvent.input(getByLabelText("New meeting item description"), {
    target: { value: newItem.description },
  });
  fireEvent.click(getByText("Create new item"));

  expect(createItemForMeeting).toBeCalledTimes(1);
  expect(createItemForMeeting).toBeCalledWith(meeting.id, {
    name: newItem.name,
    description: newItem.description,
  });
});

it("Shows an error message when creating an item fails", async () => {
  const createItemForMeeting = jest.fn(async () => {
    throw new Error("An error occured");
  });
  const { queryByText, getByText } = render(
    <MeetingDocument
      meeting={mock.meeting()}
      items={[mock.item()]}
      createItemForMeeting={createItemForMeeting}
    />
  );

  fireEvent.click(getByText("Create new item"));

  await wait(() => expect(queryByText("An error occured")).not.toBe(null));
  expect(createItemForMeeting).toBeCalledTimes(1);
});
