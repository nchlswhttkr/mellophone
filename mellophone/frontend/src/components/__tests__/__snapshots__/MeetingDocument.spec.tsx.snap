// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches a snapshot from converting an item into an element 1`] = `
<div>
  <h3>
    Item MQ==
  </h3>
  <p>
    <React.Fragment>
      Line 1
    </React.Fragment>
    <React.Fragment>
      <br />
      Line 2
    </React.Fragment>
    <React.Fragment>
      <br />
      Line 3
    </React.Fragment>
  </p>
</div>
`;
